plugins/woocommerce/src/Internal/Utilities/URL.php

```diff
@@ -137,19 +147,43 @@ private function process_path() {
 			// Directory traversals created with percent-encoding syntax should also be detected.
 			$is_traversal = str_ireplace( '%2e', '.', $part ) === '..';
 
-			// Unwind directory traversals.
-			if ( $is_traversal && count( $this->path_parts ) > 0 ) {
-				$this->path_parts = array_slice( $this->path_parts, 0, count( $this->path_parts ) - 1 );
-				continue;
+			// Resolve directory traversals (if allowed: see further comment relating to this).
+			if ( $resolve_traversals && $is_traversal ) {
+				if ( count( $this->path_parts ) > 0 && ! $retain_traversals ) {
```

Comment:
there might be an edge case when the path starts from `./`, for example, `./.././../foo.txt`

Reply:
what do we do when the path starts from other special chars, such as `~`, `-` etc?

Reply:
This note doesn't necessarily require a follow-up, I just thought there was a chance it may be handy to record a note someplace: though of course `~` at the start of a path is expanded by most shells to the current user home dir, it is not treated that way when working with `fopen()` and related functions: in those cases, it is just another symbol (for instance, if the current working dir is `/tmp`, then `~/file.txt` would mean `/tmp/~/file.txt` ... an unusual but valid path).